src/drivers/magnetometer/qmc5883/qmc5883.cpp

```diff
@@ -0,0 +1,1640 @@
+	report.y = (((int16_t)qmc_report.y[1]) << 8) + qmc_report.y[0];
+	report.z = (((int16_t)qmc_report.z[1]) << 8) + qmc_report.z[0];
+	
+
+	/*
+	 * If any of the values are -4096, there was an internal math error in the sensor.
+	 * Generalise this to a simple range check that will also catch some bit errors.
+	 */
+	if ((abs(report.x) > QMC5883_MAX_COUNT) ||
+	    (abs(report.y) > QMC5883_MAX_COUNT) ||
+	    (abs(report.z) > QMC5883_MAX_COUNT)) {
+		perf_count(_comms_errors);
+		goto out;
+	}
+
+	/* get temperature measurements from the device */
+	new_report.temperature = 0;
+
+	/*
+	 * RAW outputs
+	 *
+	 * to align the sensor axes with the board, x and y need to be flipped
+	 * and y needs to be negated
+	 */
+	new_report.x_raw = -report.y;
+	new_report.y_raw = report.x;
+	/* z remains z */
+	new_report.z_raw = report.z;
+
+	/* scale values for output */
+
+	// XXX revisit for SPI part, might require a bus type IOCTL
+	unsigned dummy;
+	sensor_is_onboard = !_interface->ioctl(MAGIOCGEXTERNAL, dummy);
+	new_report.is_external = !sensor_is_onboard;
+
+	if (sensor_is_onboard) {
+		// convert onboard so it matches offboard for the
+		// scaling below
+		report.y = -report.y;
+		report.x = -report.x;
+	}
+
+	/* the standard external mag by 3DR has x pointing to the
+	 * right, y pointing backwards, and z down, therefore switch x
+	 * and y and invert y */
+	xraw_f = -report.y;
+	yraw_f = report.x;
+	zraw_f = report.z;
+
+	// apply user specified rotation
+	rotate_3f(_rotation, xraw_f, yraw_f, zraw_f);
+	
+	new_report.x = ((xraw_f * _range_scale) - _scale.x_offset) * _scale.x_scale;
+	/* flip axes and negate value for y */
+	new_report.y = ((yraw_f * _range_scale) - _scale.y_offset) * _scale.y_scale;
+	/* z remains z */
+	new_report.z = ((zraw_f * _range_scale) - _scale.z_offset) * _scale.z_scale;
+
+	if (!(_pub_blocked)) {
+
+		if (_mag_topic != nullptr) {
+			/* publish it */
+			orb_publish(ORB_ID(sensor_mag), _mag_topic, &new_report);
+
+		} else {
+			_mag_topic = orb_advertise_multi(ORB_ID(sensor_mag), &new_report,
+							 &_orb_class_instance, (sensor_is_onboard) ? ORB_PRIO_HIGH : ORB_PRIO_MAX);
+
+			if (_mag_topic == nullptr) {
+				DEVICE_DEBUG("ADVERT FAIL");
+			}
+		}
+	}
+
+	_last_report = new_report;
+
+	/* post a report to the ring */
+	_reports->force(&new_report);
+
+	/* notify anyone waiting for data */
+	poll_notify(POLLIN);
+
+	/*
+	  periodically check the range register and configuration
+	  registers. With a bad I2C cable it is possible for the
+	  registers to become corrupt, leading to bad readings. It
+	  doesn't happen often, but given the poor cables some
+	  vehicles have it is worth checking for.
+	 */
+	check_counter = perf_event_count(_sample_perf) % 256;
+
+	if (check_counter == 0) {
+		check_range();
+	}
+
+	if (check_counter == 128) {
+		check_conf();
+	}
+
+	ret = OK;
+
+out:
+	perf_end(_sample_perf);
+	return ret;
+}
+
+int QMC5883::calibrate(struct file *filp, unsigned enable)
```

Comment:
Need to confirm that QMC5883::calibrate does anything

Reply:
This function does run when the "-C" option is given for the HMC5883.  @dagar does the calibrated values just get discarded?